File: java/leetcode/src/main/java/com/leetcode/BFS_DFS/Permutations_46.java
package com.leetcode.BFS_DFS;

import java.util.ArrayList;
import java.util.List;

public class Permutations_46 {
    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> permute(int[] nums) {
        List<Integer> path = new ArrayList<>();
        boolean[] visited = new boolean[nums.length];
        dfs(nums, 0, visited, path);
        return res;
    }

    void dfs(int[] nums, int level, boolean[] visited, List<Integer> path) {
        if (level == nums.length) {
            res.add(new ArrayList<Integer>(path));
        } else {
            for (int i = 0; i < nums.length; i++) {
                if (!visited[i]) {
                    path.add(nums[i]);
                    visited[i] = true;
                    dfs(nums, level + 1, visited, path);
                    visited[i] = false;
                    path.remove(path.size() - 1);
                }
            }
        }

    }
}

File: js/algrithm/bfs-dfs/Permutations_46.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    const visited = Array(nums.length).fill(false)
    const ans = []
    const dfs = function(level, path) {
        if (level === nums.length) {
            ans.push(path.slice())
        } else {
            for (let i = 0; i < nums.length; i++) {
                if (!visited[i]) {
                    visited[i] = true
                    path.push(nums[i])
                    dfs(level + 1, path)
                    path.pop()
                    visited[i] = false
                }
            }
        }
    }
    dfs(0, [])
    return ans 
};